CPU thread 100%

Hi,
Just updated the OS and restarted my Lemur Pro and the daemon uses 100% of one of the CPU threads.
I let it run 15 minutes then stopped the service.
Our company owns 5 identical machines, they all had the same issue with execsnoop-bpfcc

/usr/bin/system76-scheduler daemon
  \_ [execsnoop-bpfcc] < defunct >


System76 Lemur Pro/Lemur Pro, BIOS 2021-07-20_93c2809 07/20/2021


NAME="Pop!_OS"
VERSION="22.04 LTS"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 22.04 LTS"
VERSION_ID="22.04"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=jammy
UBUNTU_CODENAME=jammy
LOGO=distributor-logo-pop-os

No more defunct subprocess, but indeed execsnoop-bpfcc fails to start.


`● com.system76.Scheduler.service - Automatically configure CPU scheduler for responsiveness on AC
     Loaded: loaded (/lib/systemd/system/com.system76.Scheduler.service; disabled; vendor preset: enabled)
     Active: active (running) since Tue 2022-05-24 13:27:59 EDT; 25s ago
   Main PID: 91883 (system76-schedu)
      Tasks: 3 (limit: 149999)
     Memory: 1.6M
        CPU: 1.662s
     CGroup: /system.slice/com.system76.Scheduler.service
             └─91883 /usr/bin/system76-scheduler daemon

May 24 13:27:59 fbriand.opale.net systemd[1]: Starting Automatically configure CPU scheduler for responsiveness on AC...
May 24 13:27:59 fbriand.opale.net system76-scheduler[91883]:  INFO system76_scheduler: starting daemon service
May 24 13:27:59 fbriand.opale.net systemd[1]: Started Automatically configure CPU scheduler for responsiveness on AC.
May 24 13:27:59 fbriand.opale.net system76-scheduler[91883]:  INFO system76_scheduler::config: /usr/share/system76-scheduler/assignments/default.ron:
May 24 13:27:59 fbriand.opale.net system76-scheduler[91883]:         (Assignment(CpuPriority(-9), BestEffort(PriorityLevel(0))), {"easyeffects"})
May 24 13:27:59 fbriand.opale.net system76-scheduler[91883]:         (Assignment(CpuPriority(-5), BestEffort(PriorityLevel(4))), {"Xorg", "amsynth", "gnome-shell", "kwin", "steam", "sway", >
May 24 13:27:59 fbriand.opale.net system76-scheduler[91883]:         (Assignment(CpuPriority(0), Standard), {"dbus", "sshd", "systemd"})
May 24 13:27:59 fbriand.opale.net system76-scheduler[91883]:         (Assignment(CpuPriority(9), Idle), {"ModemManager", "NetworkManager", "accounts-daemon", "acpid", "automount", "avahi-da>
May 24 13:27:59 fbriand.opale.net system76-scheduler[91883]:         (Assignment(CpuPriority(19), Idle), {"FAHClient", "FAHCoreWrapper", "apt", "apt-get", "boinc", "c++", "cargo", "clang", >
 `



` execsnoop-bpfcc 

In file included from <built-in>:2:
In file included from /virtual/include/bcc/bpf.h:12:
In file included from include/linux/types.h:6:
In file included from include/uapi/linux/types.h:14:
In file included from include/uapi/linux/posix_types.h:5:
In file included from include/linux/stddef.h:5:
In file included from include/uapi/linux/stddef.h:5:
In file included from include/linux/compiler_types.h:80:
include/linux/compiler-clang.h:41:9: warning: '__HAVE_BUILTIN_BSWAP32__' macro redefined [-Wmacro-redefined]
#define __HAVE_BUILTIN_BSWAP32__
        ^
<command line>:4:9: note: previous definition is here
#define __HAVE_BUILTIN_BSWAP32__ 1
        ^
In file included from <built-in>:2:
In file included from /virtual/include/bcc/bpf.h:12:
In file included from include/linux/types.h:6:
In file included from include/uapi/linux/types.h:14:
In file included from include/uapi/linux/posix_types.h:5:
In file included from include/linux/stddef.h:5:
In file included from include/uapi/linux/stddef.h:5:
In file included from include/linux/compiler_types.h:80:
include/linux/compiler-clang.h:42:9: warning: '__HAVE_BUILTIN_BSWAP64__' macro redefined [-Wmacro-redefined]
#define __HAVE_BUILTIN_BSWAP64__
        ^
<command line>:5:9: note: previous definition is here
#define __HAVE_BUILTIN_BSWAP64__ 1
        ^
In file included from <built-in>:2:
In file included from /virtual/include/bcc/bpf.h:12:
In file included from include/linux/types.h:6:
In file included from include/uapi/linux/types.h:14:
In file included from include/uapi/linux/posix_types.h:5:
In file included from include/linux/stddef.h:5:
In file included from include/uapi/linux/stddef.h:5:
In file included from include/linux/compiler_types.h:80:
include/linux/compiler-clang.h:43:9: warning: '__HAVE_BUILTIN_BSWAP16__' macro redefined [-Wmacro-redefined]
#define __HAVE_BUILTIN_BSWAP16__
        ^
<command line>:3:9: note: previous definition is here
#define __HAVE_BUILTIN_BSWAP16__ 1
        ^
3 warnings generated.
cannot attach kprobe, probe entry may not exist
Traceback (most recent call last):
  File "/usr/sbin/execsnoop-bpfcc", line 229, in <module>
    b.attach_kprobe(event=execve_fnname, fn_name="syscall__execve")
  File "/usr/lib/python3/dist-packages/bcc/__init__.py", line 683, in attach_kprobe
    raise Exception("Failed to attach BPF program %s to kprobe %s" %
Exception: Failed to attach BPF program b'syscall__execve' to kprobe b'sys_execve'`

--- a/daemon/src/config.rs
+++ b/daemon/src/config.rs
@@ -101,13 +101,26 @@ impl From<u8> for PriorityLevel {
     }
 }
 
-#[derive(Debug, Default, Deserialize, Serialize)]
+#[derive(Debug, Deserialize, Serialize)]
 pub struct Config {
     #[serde(default)]
     pub background: Option<i8>,
 
     #[serde(default)]
     pub foreground: Option<i8>,
+
+    #[serde(default)]
+    pub use_execsnoop: bool,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            background: Some(5),
+            foreground: Some(-5),
+            use_execsnoop: true,
+        }
+    }
 }
 
 impl Config {
@@ -138,10 +151,7 @@ impl Config {
 
         tracing::info!("Using default config values due to config error");
 
-        Config {
-            background: Some(5),
-            foreground: Some(-5),
-        }
+        Config::default()
     }
 }
 

--- a/data/config.ron
+++ b/data/config.ron
@@ -7,4 +7,7 @@ background: Some(5),
 
 // The priority to assign foreground tasks.
 foreground: Some(-5),
+
+// Enables realtime process priority adjustment.
+use_execsnoop: true,
 )

--- a/execsnoop/Cargo.toml
+++ b/execsnoop/Cargo.toml
@@ -8,4 +8,3 @@ license = "MPL-2.0"
 publish = true
 
 [dependencies]
-compact_str = "0.3.2"

--- a/execsnoop/src/lib.rs
+++ b/execsnoop/src/lib.rs
@@ -4,7 +4,7 @@
 use std::{
     io,
     io::{BufRead, BufReader},
-    process::{Command, Stdio}, iter::FromFn,
+    process::{Command, Stdio},
 };
 
 #[derive(Clone, Debug)]
@@ -15,7 +15,7 @@ pub struct Process {
 
 pub struct ProcessIterator {
     child: std::process::Child,
-    iterator: Box<dyn Iterator<Item = Process>>
+    iterator: Box<dyn Iterator<Item = Process> + Send>,
 }
 
 impl Iterator for ProcessIterator {
@@ -38,7 +38,7 @@ impl Drop for ProcessIterator {
 /// # Errors
 ///
 /// Requires the `execsnoop-bpfcc` binary from `bpfcc-tools`
-pub fn watch() -> io::Result<impl Iterator<Item = Process>> {
+pub fn watch() -> io::Result<ProcessIterator> {
     Command::new(std::env!(
         "EXECSNOOP_PATH",
         "must set EXECSNOOP_PATH env to execsnoop-bpfcc path"
@@ -50,6 +50,8 @@ pub fn watch() -> io::Result<impl Iterator<Item = Process>> {
     .spawn()
     .and_then(|mut child| {
         let stdout = child.stdout.take().ok_or_else(|| {
+            let _res = child.kill();
+            let _res = child.wait();
             io::Error::new(io::ErrorKind::Other, "execsnoop-bpfcc lacks stdout pipe")
         })?;
 
@@ -61,9 +63,7 @@ pub fn watch() -> io::Result<impl Iterator<Item = Process>> {
             child,
             iterator: Box::new(std::iter::from_fn(move || loop {
                 match reader.read_line(&mut line) {
-                    Err(_) | Ok(0) => {
-                        return None
-                    },
+                    Err(_) | Ok(0) => return None,
                     _ => (),
                 }
 
@@ -85,7 +85,7 @@ pub fn watch() -> io::Result<impl Iterator<Item = Process>> {
                 }
 
                 line.clear();
-            }))
+            })),
         })
     })
 }

--- a/execsnoop/src/lib.rs
+++ b/execsnoop/src/lib.rs
@@ -36,29 +36,30 @@ pub fn watch() -> io::Result<impl Iterator<Item = Process>> {
 
         let mut line = String::with_capacity(128);
 
-        Ok(std::iter::from_fn(move || {
-            while reader.read_line(&mut line).is_ok() {
-                let mut fields = line.split_ascii_whitespace();
-
-                let pid = fields.nth(1);
-                let parent_pid = fields.next();
-
-                if let Some((pid, parent_pid)) = pid.zip(parent_pid) {
-                    let pid = pid.parse::<u32>().ok();
-                    let parent_pid = parent_pid.parse::<u32>().ok();
-
-                    if let Some((pid, parent_pid)) = pid.zip(parent_pid) {
-                        let process = Process { pid, parent_pid };
-
-                        line.clear();
-                        return Some(process);
-                    }
-                }
-
-                line.clear();
+        Ok(std::iter::from_fn(move || loop {
+            match reader.read_line(&mut line) {
+                Err(_) | Ok(0) => return None,
+                _ => (),
             }
 
-            None
+            let mut fields = line.split_ascii_whitespace();
+
+            let pid = fields.nth(1);
+            let parent_pid = fields.next();
+
+            if let Some((pid, parent_pid)) = pid.zip(parent_pid) {
+                let pid = pid.parse::<u32>().ok();
+                let parent_pid = parent_pid.parse::<u32>().ok();
+
+                if let Some((pid, parent_pid)) = pid.zip(parent_pid) {
+                    let process = Process { pid, parent_pid };
+
+                    line.clear();
+                    return Some(process);
+                }
+            }
+
+            line.clear();
         }))
     })
 }

--- a/execsnoop/src/lib.rs
+++ b/execsnoop/src/lib.rs
@@ -4,13 +4,33 @@
 use std::{
     io,
     io::{BufRead, BufReader},
-    process::{Command, Stdio},
+    process::{Command, Stdio}, iter::FromFn,
 };
 
 #[derive(Clone, Debug)]
 pub struct Process {
     pub pid: u32,
     pub parent_pid: u32,
+}
+
+pub struct ProcessIterator {
+    child: std::process::Child,
+    iterator: Box<dyn Iterator<Item = Process>>
+}
+
+impl Iterator for ProcessIterator {
+    type Item = Process;
+
+    fn next(&mut self) -> Option<Process> {
+        self.iterator.next()
+    }
+}
+
+impl Drop for ProcessIterator {
+    fn drop(&mut self) {
+        let _res = self.child.kill();
+        let _res = self.child.wait();
+    }
 }
 
 /// Watches process creation and destruction on Linux.
@@ -37,30 +57,35 @@ pub fn watch() -> io::Result<impl Iterator<Item = Process>> {
 
         let mut line = String::with_capacity(128);
 
-        Ok(std::iter::from_fn(move || loop {
-            match reader.read_line(&mut line) {
-                Err(_) | Ok(0) => return None,
-                _ => (),
-            }
+        Ok(ProcessIterator {
+            child,
+            iterator: Box::new(std::iter::from_fn(move || loop {
+                match reader.read_line(&mut line) {
+                    Err(_) | Ok(0) => {
+                        return None
+                    },
+                    _ => (),
+                }
 
-            let mut fields = line.split_ascii_whitespace();
+                let mut fields = line.split_ascii_whitespace();
 
-            let pid = fields.nth(1);
-            let parent_pid = fields.next();
-
-            if let Some((pid, parent_pid)) = pid.zip(parent_pid) {
-                let pid = pid.parse::<u32>().ok();
-                let parent_pid = parent_pid.parse::<u32>().ok();
+                let pid = fields.nth(1);
+                let parent_pid = fields.next();
 
                 if let Some((pid, parent_pid)) = pid.zip(parent_pid) {
-                    let process = Process { pid, parent_pid };
+                    let pid = pid.parse::<u32>().ok();
+                    let parent_pid = parent_pid.parse::<u32>().ok();
 
-                    line.clear();
-                    return Some(process);
+                    if let Some((pid, parent_pid)) = pid.zip(parent_pid) {
+                        let process = Process { pid, parent_pid };
+
+                        line.clear();
+                        return Some(process);
+                    }
                 }
-            }
 
-            line.clear();
-        }))
+                line.clear();
+            }))
+        })
     })
 }

--- a/execsnoop/src/lib.rs
+++ b/execsnoop/src/lib.rs
@@ -23,6 +23,7 @@ pub fn watch() -> io::Result<impl Iterator<Item = Process>> {
         "EXECSNOOP_PATH",
         "must set EXECSNOOP_PATH env to execsnoop-bpfcc path"
     ))
+    .env("LC_ALL", "C")
     .stdout(Stdio::piped())
     .stderr(Stdio::null())
     .stdin(Stdio::null())

--- a/execsnoop/src/lib.rs
+++ b/execsnoop/src/lib.rs
@@ -1,7 +1,6 @@
 // Copyright 2022 System76 <info@system76.com>
 // SPDX-License-Identifier: MPL-2.0
 
-use compact_str::CompactStr;
 use std::{
     io,
     io::{BufRead, BufReader},
@@ -10,7 +9,6 @@ use std::{
 
 #[derive(Clone, Debug)]
 pub struct Process {
-    pub comm: CompactStr,
     pub pid: u32,
     pub parent_pid: u32,
 }
@@ -42,20 +40,15 @@ pub fn watch() -> io::Result<impl Iterator<Item = Process>> {
             while reader.read_line(&mut line).is_ok() {
                 let mut fields = line.split_ascii_whitespace();
 
-                let command = fields.next();
-                let pid = fields.next();
+                let pid = fields.nth(1);
                 let parent_pid = fields.next();
 
-                if let Some(((command, pid), parent_pid)) = command.zip(pid).zip(parent_pid) {
+                if let Some((pid, parent_pid)) = pid.zip(parent_pid) {
                     let pid = pid.parse::<u32>().ok();
                     let parent_pid = parent_pid.parse::<u32>().ok();
 
                     if let Some((pid, parent_pid)) = pid.zip(parent_pid) {
-                        let process = Process {
-                            comm: CompactStr::new(command),
-                            pid,
-                            parent_pid,
-                        };
+                        let process = Process { pid, parent_pid };
 
                         line.clear();
                         return Some(process);

